Fix ddp loderunner

hickmank opened the pull request:
I have modified the non-HDF5 save and load functions so that the optimizer is always attached to the model each rank is actually trying to update. 

Prior to this the `ch_DDP_loderunner` harness would stop updating the model after the first checkpoint.

galegozi commented:
Hi @hickmank ,
I will look at the changes, but given the testing failed, we will need to re-review it once all tests pass! 
Thanks! 
Gal

galegozi commented:
> > Hi @hickmank , I will look at the changes, but given the testing failed, we will need to re-review it once all tests pass! Thanks! Gal
> 
> Yes, the test was failing since the non-HDF5 checkpointing now returns an optimizer from the load function. The load function was also changed so that an optimizer class and kwargs are passed in instead of an optimizer. This is necessary since the load actually instantiates a new model when it is called. Therefore it must instantiate an optimizer and attach the weights of the new model to that optimizer. Prior to this PR the checkpointed model and optimizer were left in a state where the parameters that the optimizer was updating were not the ones in the model that was being trained. I've just updated the unit tests to reflect these modifications.

Excellent, thanks! I will formally approve once all the testing passes, just hope that it will not cause too much issues for the other harnesses that rely on this :)

hickmank commented:
> > > Hi @hickmank , I will look at the changes, but given the testing failed, we will need to re-review it once all tests pass! Thanks! Gal
> > 
> > 
> > Yes, the test was failing since the non-HDF5 checkpointing now returns an optimizer from the load function. The load function was also changed so that an optimizer class and kwargs are passed in instead of an optimizer. This is necessary since the load actually instantiates a new model when it is called. Therefore it must instantiate an optimizer and attach the weights of the new model to that optimizer. Prior to this PR the checkpointed model and optimizer were left in a state where the parameters that the optimizer was updating were not the ones in the model that was being trained. I've just updated the unit tests to reflect these modifications.
> 
> Excellent, thanks! I will formally approve once all the testing passes, just hope that it will not cause too much issues for the other harnesses that rely on this :)

The harnesses that use this will break. As they should. @Smyther-Hsiao @ryanw18 @soumide1102 @wshlavacek will be affected by this. However, the `ch_DDP_loderunner` training script has a pretty clear map for what SHOULD be done.

galegozi reviewed (approved):
I am approving, and excited to see the results, thanks!
